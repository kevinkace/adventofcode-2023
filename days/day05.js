export class Almanac {
    constructor(str) {
        this.str = str;
        this.maps = [];

        this.parse();
    }

    parse() {
        this.str.split(/(?:\r?\n){2}/).forEach((section, idx) => {
            if (idx === 0) {
                this.seeds = this.parseSeeds(section);

                return;
            }

            const map = {
                name : "",
                rows : []
            };

            section.split(/\r?\n/).forEach((row, idx) => {
                if (idx === 0) {
                    map.name = row;

                    return;
                }

                map.rows.push(row.split(" ").map(Number));
            });

            this.maps.push(map);
        });
    }

    toString() {
        const seedsStr = this.seedsToString();
        const maps = Almanac.mapsToString(this.maps);

        return [seedsStr, ...maps].join("\r\n\r\n");
    }

    parseSeeds(section) {
        return section.split(" ").filter((_, idx) => idx).map(Number);
    }

    seedsToString() {
        return `seeds: ${this.seeds.join(" ")}`;
    }

    static mapsToString(maps) {
        return maps.map((map) => {
            const rows = map.rows.map((row) => row.join(" ")).join("\r\n");

            return `${map.name}\r\n${rows}`;
        });
    }

    _findMap(from, map) {
        return Almanac._findMap(from, map);
    }

    _getDestFromMap(from, map) {
        return Almanac._getDestFromMap(from, map);
    }

    _getDest(from, map) {
        return Almanac._getDest(from, map);
    }

    _getFinalDestFromMaps(from, maps) {
        return Almanac._getFinalDestFromMaps(from, maps);
    }

    static _findMap(from, maps) {
        return maps.find((row, idx) => {
            const [ dest, src, range ] = row;

            return src <= from && from <= src + range;
        });
    }

    static _getDestFromMap(from, map) {
        return map[0] + (from - map[1]);
    }

    static _getDest(from, map) {
        const foundMap = this._findMap(from, map);

        return foundMap ? this._getDestFromMap(from, foundMap) : from;
    }

    static _getFinalDestFromMaps(from, maps) {
        return maps.reduce((acc, map) => {
            const dest = this._getDest(acc, map);

            // console.log({ acc, map, dest });

            return dest;
        }, from);
    }

    getFinalDestForAllSeeds() {
        return this.seeds.map((seed) => this._getFinalDestFromMaps(seed, this.maps.map((map) => map.rows)));
    }

    getMinFinalDestForAllSeeds() {
        return Math.min(...this.getFinalDestForAllSeeds());
    }
}


export class Almanac2 extends Almanac {
    constructor(str) {
        super(str);
    }

    parseSeeds(sections) {
        return sections
            .split(" ")
            .filter((_, idx) => idx) // remove first el `"seeds:"`
            .map(Number)
            // group in touples
            .reduce((acc, seed, idx) => {
                const pos = Math.floor(idx / 2);

                if (!acc[pos]) {
                    acc[pos] = [];
                }

                acc[pos].push(seed);

                return acc;
            }, []);
    }

    seedsToString() {
        return `seeds: ${this.seeds.flat().join(" ")}`;
    }

    static _getRangesFromRange(inputRange, ranges) {
        let [ start, length ] = inputRange;
        const end = start + length;

        const outputRanges = this._filterRangesBetween(start, end, ranges);

        if (!outputRanges.length) {
            return [[ start, start, length ]];
        }

        const firstRange = outputRanges.at(0);
        const lastRange = outputRanges.at(-1);
        const lastRangeEnd = lastRange[1] + lastRange[2];

        if (firstRange[1] > start) {
            outputRanges.unshift([ start, start, firstRange[1] - start ]);
        }

        if (lastRangeEnd < end) {
            outputRanges.push([ lastRangeEnd, lastRangeEnd, end - lastRangeEnd ]);
        }

        return outputRanges;
    }

    static _getDestRangesFromOutRanges(inputRange, ranges) {
        const [ start, length ] = inputRange;
        const end = start + length;

        const outRanges = this._getRangesFromRange(inputRange, ranges);

        return outRanges.map(([ dest, src, mapRange ]) => {
            const outRangeDestStart  = dest + (start - src);
            const outRangeDestRange = mapRange - (start - src) - (src + mapRange - end);

            return [ outRangeDestStart, outRangeDestRange ];
        });
    }

    static _filterRangesBetween(start, end, ranges) {
        return Almanac2._sortMapsBySrc(ranges)
        .filter(([ dest, src, mapRange ], idx) => {

            const srcEnd = src + mapRange;

            if (start >= src && start <= srcEnd) {
                return true;
            }

            if (end >= src && end <= srcEnd) {
                return true;
            }

            return false;
        });
    }

    static _sortMapsBySrc(maps) {
        return maps.sort((a, b) => a[1] - b[1]);
    }

}