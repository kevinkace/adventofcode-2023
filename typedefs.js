/**
 * @typedef {Object} Game
 * @property {number} game
 * @property {Round[]} rounds
 */

/**
 * @typedef {Object} Round
 * @property {number} round
 * @property {number} red
 * @property {number} green
 * @property {number} blue
 */

export default {};
