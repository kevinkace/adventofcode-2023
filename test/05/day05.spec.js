import { readFile }      from "node:fs/promises";
import { fileURLToPath } from "node:url";
import path              from "node:path";

import { expect, test, describe } from "vitest";

import { Day05 } from "../../days/day05.js";

import typedefs from "../../typedefs.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname  = path.dirname(__filename);

const sample = await readFile(path.join(__dirname, "sample.txt"), "utf8");
const input  = await readFile(path.join(__dirname, "input.txt"), "utf8");


describe("day 05", () => {
    test("can parse input", () => {
        expect((new Day05(sample)).toString()).toBe(sample);
        expect((new Day05(input)).toString()).toBe(input);
    });
});
