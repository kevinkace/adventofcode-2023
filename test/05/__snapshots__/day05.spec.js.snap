// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`day 05 > Almanac 2 parse 1`] = `
[
  [
    79,
    14,
  ],
  [
    55,
    13,
  ],
]
`;

exports[`day 05 > can sort maps 1`] = `
[
  {
    "outRanges": [
      [
        52,
        50,
        48,
      ],
    ],
    "seed": [
      79,
      14,
    ],
  },
  {
    "outRanges": [
      [
        79,
        79,
        14,
      ],
    ],
    "seed": [
      79,
      14,
    ],
  },
  {
    "outRanges": [
      [
        79,
        79,
        14,
      ],
    ],
    "seed": [
      79,
      14,
    ],
  },
  {
    "outRanges": [
      [
        18,
        25,
        70,
      ],
    ],
    "seed": [
      79,
      14,
    ],
  },
  {
    "outRanges": [
      [
        45,
        77,
        23,
      ],
    ],
    "seed": [
      79,
      14,
    ],
  },
  {
    "outRanges": [
      [
        79,
        79,
        14,
      ],
    ],
    "seed": [
      79,
      14,
    ],
  },
  {
    "outRanges": [
      [
        60,
        56,
        37,
      ],
      [
        56,
        93,
        4,
      ],
    ],
    "seed": [
      79,
      14,
    ],
  },
  {
    "outRanges": [
      [
        52,
        50,
        48,
      ],
    ],
    "seed": [
      55,
      13,
    ],
  },
  {
    "outRanges": [
      [
        55,
        55,
        13,
      ],
    ],
    "seed": [
      55,
      13,
    ],
  },
  {
    "outRanges": [
      [
        49,
        53,
        8,
      ],
      [
        61,
        61,
        7,
      ],
    ],
    "seed": [
      55,
      13,
    ],
  },
  {
    "outRanges": [
      [
        18,
        25,
        70,
      ],
    ],
    "seed": [
      55,
      13,
    ],
  },
  {
    "outRanges": [
      [
        81,
        45,
        19,
      ],
      [
        68,
        64,
        13,
      ],
    ],
    "seed": [
      55,
      13,
    ],
  },
  {
    "outRanges": [
      [
        1,
        0,
        69,
      ],
    ],
    "seed": [
      55,
      13,
    ],
  },
  {
    "outRanges": [
      [
        55,
        55,
        1,
      ],
      [
        60,
        56,
        37,
      ],
    ],
    "seed": [
      55,
      13,
    ],
  },
]
`;

exports[`day 05 > expand seed touples 1`] = `
[
  79,
  80,
  81,
  82,
  83,
  84,
  85,
  86,
  87,
  88,
  89,
  90,
  91,
  92,
  55,
  56,
  57,
  58,
  59,
  60,
  61,
  62,
  63,
  64,
  65,
  66,
  67,
]
`;
